planner/core/rule_topn_push_down.go

```diff
@@ -147,7 +147,7 @@ func (p *LogicalJoin) pushDownTopNToChild(topN *LogicalTopN, idx int) LogicalPla
 	for _, by := range topN.ByItems {
 		cols := expression.ExtractColumns(by.Expr)
 		for _, col := range cols {
-			if p.children[1-idx].Schema().Contains(col) {
+			if p.children[1-idx].Schema().Contains(col) || col.ID == 0 /* contains non-table column */ {
```

Comment:
or maybe we should :rofl: 

```
!p.children[idx].Schema().Contains(col)
```

Reply:
This one is better. You can try a case that the join is `select * from (select a+b as x from t1) t1, t2 where t1.x=t2.a` and topn's order by column is `t1.x`.

Here `t1.x`'s `ID` is `0` since it's not a column from actual table.